### ISSUE TITLE
Restore doesn't recurse into directories

### ISSUE THREAD

**Issue by Germar** [bit-team/backintime#363]
When Firefox4 crashed and nuked my profile, I looked to BiT to help.

I moved ~/.mozilla out of the way and clicked "restore" on that directory in BiT's snapshot, but it just created the empty directory.

I then clicked "restore" on the directories inside that - firefox, plugins, eclipse, extensions; but when I launched FF4 again my extensions weren't there.

It turns out that SOME subdirectories of those directories weren't restored either!

As some directories are and some directories aren't recursed into, I assume this is a bug.

1.0.6 on Fedora 13.

------------------------------------
Imported from Launchpad using lp2gh.

 * date created: 2011-05-12T08:40:30Z
 * owner: sej7278
 * the launchpad url was https://bugs.launchpad.net/bugs/781550

**Comment by Germar** [comment]
(by sej7278)
Running from the console, I can see the error is down to permissions, but I wasn't warned about that when I took the snapshot.

And why is there a permission issue on a 700 directory when the rsync process is running as my user? Do we really have to run it as root to backup one user's directories?

INFO: Restore: /home/user/.mozilla
rsync: recv_generator: mkdir "/home/user/.mozilla/eclipse" failed: Permission denied (13)
*** Skipping any contents from this failed directory ***
rsync: recv_generator: mkdir "/home/user/.mozilla/extensions" failed: Permission denied (13)
*** Skipping any contents from this failed directory ***
rsync: recv_generator: mkdir "/home/user/.mozilla/firefox" failed: Permission denied (13)
*** Skipping any contents from this failed directory ***
rsync: recv_generator: mkdir "/home/user/.mozilla/plugins" failed: Permission denied (13)
*** Skipping any contents from this failed directory ***
rsync error: some files/attrs were not transferred (see previous errors) (code 23) at main.c(1052) [sender=3.0.8]
WARNING: Command "rsync -rtDH --links --no-p --no-g --no-o -R --backup --suffix=.backup.20110512 "/backups/server/user/1/20110511-172859-448/backup/./home/user/.mozilla" /" returns 5888
INFO: Restore permissions

**Comment by Germar** [comment]
(by trent-trentonadams)
Are you using se-linux?

**Comment by Germar** [comment]
(by trent-trentonadams)
I just tested this on my computer, by setting my folder that I'm restoring to, to 700 permissions, and it works as expected.

**Comment by Germar** [comment]
(by sej7278)
selinux is disabled

my $HOME directory is set to 700 perms, so i guess that's the problem - backing up and restoring as root works fine.

**Comment by Germar** [comment]
(by lopo)
Looks like a duplicate of https://bugs.launchpad.net/backintime/+bug/813462

**Comment by Germar** [comment]
(by sej7278)
my bug was here first, so the other one is a duplicate of this.

essentially rsync needs to run as root to be able to change permissions, therefore backintime should warn you at least that its not going to be able to restore properly, rather than silently failing.

these days i just use the "Back In Time (root)" launcher which uses gksu to make and restore backups as root.

1.0.7 on debian wheezy.

**Comment by Germar** [comment]
(by mgiuca)
Marked bug #813462 and bug #788280 as duplicates of this one. Since bug #813462 was marked confirmed, I suggest changing this bug's status to confirmed as well.

**Comment by Germar** [comment]
(by mgiuca)
Oh, I can change it to confirmed -- I'll do that.

**Comment by Germar** [comment]
(by mgiuca)
I've fixed this bug -- can someone test to see that it worked? I have the patch in the branch lp:~mgiuca/backintime/restore-chmod. There is a merge request: https://code.launchpad.net/~mgiuca/backintime/restore-chmod/+merge/82814

The correct solution is *NOT* to run rsync as root. That would be very dangerous. (Therefore, running Back In Time (root) is a workaround, but a dangerous one.)

The problem is that rsync is copying directories which do not have write permission. Therefore, if you have a directory called "foo" with a file "bar" in it, it copies the directory "foo" to the restore destination (setting its permissions to 555 or read-only), and then attempts to copy "bar", but it can't, because "foo" is read-only.

The correct solution is to tell rsync to set each directory's permissions to 777 while it is copying. This can be done by passing --chmod=ugo=rwX to rsync (see man rsync, under --chmod). This guarantees that it is writable when rsync copies the contents. The fact that we are blasting away the permissions is not a problem -- once rsync is finished, BIT is going to correctly set the permissions for each file, so in the end all of the permissions will be set correctly.

**Comment by Germar** [comment]
(by danleweb)
I merged the brunch.

Thanks,
Dan

**Comment by Germar** [comment]
(by anrichter)
Would there be a new version with the fix included? I have installed backintime 1.0.8-1 on ubuntu 11.10 and the bug still exists.

**Comment by Germar** [comment]
(by danleweb)
Not yet ... the next version will be 1.0.10.

**Comment by Germar** [comment]
(by mashimom)
Hi, since the fix was in by November, when will 1.0.10 build arrive? Should I build from source myself to get it? The root execution did not show my backups, so as far as workaround I have none.
